src/planet-a/PlanetGeometryA.cpp

```diff
@@ -352,70 +370,99 @@ void PlanetGeometryA::chunk_add(trindex_t t)
         {
             // a shared triangle can be taken from a corner
-            vertIndex = sharedCorner;
+            vertIndex = rSharedCorner;
             m_vrtxSharedUsers[vertIndex] ++;
-            std::cout << "Corner taken!\n";
         }
-        else
+
+        if (vertIndex == gc_invalidVrtx)
         {
-            bool dummy = false;
 
             // Get a vertex from either of the sides
 
-            vertIndex  = shared_from_neighbour(t, sideA, m_chunkWidthB, dummy);
+            // try side A
+            vertIndex = shared_from_neighbour(triInd, sideA, m_chunkWidthB);
 
+            // if that fails, try side B
             if (vertIndex == gc_invalidVrtx)
             {
-                vertIndex  = shared_from_neighbour(t, sideB, 0, dummy);
+                vertIndex = shared_from_neighbour(triInd, sideB, 0);
             }
 
-            if (vertIndex == gc_invalidVrtx)
-            {
-                vertIndex = shared_create();
-            }
-            else
+            // success, increment user count and stuff
+            if (vertIndex != gc_invalidVrtx)
             {
                 m_vrtxSharedUsers[vertIndex] ++;
+
+                rSharedCorner = vertIndex;
+                m_vrtxSharedIcoCornersReverse[vertIndex] = triCorner;
             }
-            sharedCorner = vertIndex;
         }
 
-        auto vrtxOffset = m_vrtxBuffer.begin() + vertIndex * m_vrtxSize;
+        // if all sharing attempts fail, then create a new one
+        if (vertIndex == gc_invalidVrtx)
+        {
+            vertIndex = shared_create();
+
+            rSharedCorner = vertIndex;
+            m_vrtxSharedIcoCornersReverse[vertIndex] = triCorner;
+        }
+
+
+        buindex_t vrtxOffset = vertIndex * m_vrtxSize;
+        auto vrtxOffsetIt = m_vrtxBuffer.begin() + vrtxOffset;
 
         // sideB because local ringed indices start at the bottom left
         indices[sideB * m_chunkWidthB] = vertIndex;
         initTri[corner].m_vrtxIndex = vertIndex;
 
-        float *vrtxDataRead = m_icoTree->get_vertex_pos(tri.m_corners[corner]);
+        //float *vrtxDataRead = m_icoTree->get_vertex_pos(tri.m_corners[corner]);
+        vrindex_t icoCornerVrtx = rTri.m_corners[corner];
 
         // Copy Icotree's position and normal data to vertex buffer
 
-        std::copy(vrtxDataRead + m_icoTree->m_vrtxCompOffsetPos,
-                  vrtxDataRead + m_icoTree->m_vrtxCompOffsetPos + 3,
-                  vrtxOffset + m_vrtxCompOffsetPos);
+        Vector3 const& vrtxPos = Vector3::from(
+                    m_icoTree->get_vertex_pos(icoCornerVrtx));
+
+        Vector3 const& vrtxNrm = Vector3::from(
+                    m_icoTree->get_vertex_nrm(icoCornerVrtx));
+
+        Vector3 pos = vrtxPos;
+        Vector3 nrm = pos.normalized();
+
+        // temporary
+        pos = nrm * float(m_icoTree->get_radius());
+        pos += nrm * debug_stupid_heightmap(pos);
+
+        std::copy(pos.data(), pos.data() + pos.Size,
```

Comment:
not pos.begin(), pos.end() ?

Reply:
These are Vector3, they don't have iterators.

Reply:
Gross. Alrighty.